lib/charms/layer/ruby.py

```diff
@@ -1,111 +1,76 @@
-    sh = shell(shell_cmd)
-    if sh.code > 0:
-        hookenv.status_set('blocked',
-                           'Problem with Ruby: {}'.format(sh.errors()))
+    is_snap_install = config.get("ruby-snap", None)
+    ruby_install_version = config.get("ruby-install-version", None)
+    if is_snap_install and not ruby_install_version:
+        sh.snap.install("ruby --channel {} --classic".format(config.get("ruby-snap")))
+        hookenv.status_set("maintenance", "Ruby installed from snapstore.")
+        return
+
+    stem = "ruby-install-{}".format(ruby_install_version)
+    tarfile = "ruby-install-{}.tar.gz".format(ruby_install_version)
+    tarfile_gpg = "ruby-install-{}.tar.gz.asc".format(ruby_install_version)
+    sh.wget(
+        "-O",
+        "/tmp/postmodern.asc",
+        "https://raw.github.com/postmodern/postmodern.github.io/master/postmodern.asc",
+    )
+    sh.gpg("--import", "/tmp/postmodern.asc")
+    sh.wget(
+        "-O",
+        "/tmp/{}".format(tarfile),
+        "https://github.com/postmodern/ruby-install/archive/v{}.tar.gz".format(
+            ruby_install_version
+        ),
+    )
+    sh.wget(
+        "-O",
+        "/tmp/{}".format(tarfile_gpg),
+        "https://raw.github.com/postmodern/ruby-install/master/pkg/{}".format(
+            tarfile_gpg
+        ),
+    )
+    sh.tar("-xzvf", tarfile, _cwd="/tmp")
+
+    for line in sh.make("install", _cwd=stem, _iter=True, _bg_exc=False):
+        sh.echo(line)
+
+    try:
+        sh.gpg("--verify", "/tmp/{}".format(tarfile_gpg), "/tmp/{}".format(tarfile))
+    except sh.ErrorReturnCode as error:
+        hookenv.status_set("blocked", "Problem getting ruby-install: {}".format(error))
+        hookenv.log("Problem getting ruby-install: {}".format(error))
         sys.exit(1)
 
 
-def tarball_exists(url):
-    """ Checks that ruby tarball exists on mirror
-    """
-    resp = requests.head(url)
-    if resp.status_code == 200:
-        return True
-    return False
-
-
 def compile_ruby():
-    with chdir(WORKDIR):
-        cmds = [
-            'env RUBY_CFLAGS="-O3" ./configure --prefix=/usr '
-            '--disable-install-rdoc',
-            'make -j{}'.format(cpu_count()),
-            'make install'
-        ]
-
-        for cmd in cmds:
-            hookenv.log('Running compile command: {}'.format(cmd))
-            sh = shell(cmd, record_output=False)
-            if sh.code > 0:
-                hookenv.status_set('blocked',
-                                   'Problem with Ruby: {}'.format(sh.errors()))
-                hookenv.log("Problem with Ruby: {}".format(sh.errors()))
-                sys.exit(1)
-    hookenv.status_set('maintenance', 'Installing Ruby completed.')
-
-
-def download_ruby():
-    """ Downloads ruby tarball, puts it in /tmp
-    """
-    url = os.path.join(config['ruby-mirror'],
-                       'ruby-{}.tar.gz'.format(config['ruby-version']))
-    if not tarball_exists(url):
+    version = config.get("ruby-version", "2.6.4")
```

Comment:
This doesn't work as you expect, as the default will never be used. config['ruby-version'] always contains a value (even if None or the default '').

Instead:
version = config['ruby-version'] or '2.6.4'